File: day15/main.go
package main

import (
	"fmt"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func playGameUntil(until, turnno, cur int, num2turn map[int]int) int {
	for ; turnno < until; turnno += 1 {
		last := cur
		if lastSeen, notFirst := num2turn[cur]; notFirst {
			cur = turnno - lastSeen
		} else {
			cur = 0
		}
		num2turn[last] = turnno
	}
	return cur
}

func withFresh(until int) int {
	input := []int{0, 13, 1, 16, 6, 17}
	num2turn := map[int]int{}
	for turn, num := range input {
		num2turn[num] = turn + 1
	}

	turnno := len(input)
	start := input[turnno-1]
	return playGameUntil(until, turnno, start, num2turn)
}

func main() {
	fmt.Println("Part 1:")
	fmt.Println(withFresh(2020))

	fmt.Println("Part 2:")
	fmt.Println(withFresh(30000000))
}

File: day15/Main.hs
import           Data.List
import           Data.Map                       ( Map )
import qualified Data.Map                      as Map

stepGame :: (Int, Int, Map Int Int) -> (Int, Map Int Int)
stepGame (turnno, cur, state) =
  let state' = Map.insert cur turnno state
  in  if Map.notMember cur state
        then (0, state')
        else (turnno - Map.findWithDefault 0 cur state, state')

doGame :: Int -> Int -> Map Int Int -> Int
doGame 30000000 cur _ = cur
doGame turnno cur state =
  let (next, state') = stepGame (turnno, cur, state)
  in  doGame (turnno + 1) next state'

main = do
  let input = [0, 13, 1, 16, 6, 17]
  let num2turn =
        foldl' (\map (i, n) -> Map.insert n i map) Map.empty (zip [1 ..] input)
  print "Part 1:"
  print $ doGame (length input) (last input) num2turn
  print "Part 2:"
